diff --git a/parser.ml b/parser.ml
index d535082..a4b09f6 100644
--- a/parser.ml
+++ b/parser.ml
@@ -199,7 +199,10 @@ let parse_pretype =
           (try pretype_of_type(assoc s (type_abbrevs())) with Failure _ ->
            try mk_prefinty (num_of_string s) with Failure _ ->
            if try get_type_arity s = 0 with Failure _ -> false
-           then Ptycon(s,[]) else Utv(s)),rest
+           then Ptycon(s,[]) else (
+             if ((String.compare (String.sub s 0 (Pervasives.min (String.length s) 3)) "_?_") == 0)
+             then Utv("?"^(String.sub s 3 ((String.length s)-3))) else Utv(s)
+           )),rest
     | _ -> raise Noparse
   and type_constructor input =
     match input with
